File: app/schemas/price_desk.py
"""Pydantic-контракты API стола прайсов."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

PriceDeskKind = Literal["plates", "fbs", "march", "step", "bridge_pile", "pile"]

STATUS_KINDS: tuple[PriceDeskKind, ...] = (
    "plates",
    "fbs",
    "march",
    "step",
    "bridge_pile",
    "pile",
)


class PriceDeskDiffExample(BaseModel):
    key: str
    old_price: float | None = None
    new_price: float | None = None


class PriceDeskDiffExamples(BaseModel):
    changed: list[PriceDeskDiffExample] = Field(default_factory=list)
    new: list[PriceDeskDiffExample] = Field(default_factory=list)
    missing: list[PriceDeskDiffExample] = Field(default_factory=list)
    unchanged: list[PriceDeskDiffExample] = Field(default_factory=list)


class PriceDeskPreview(BaseModel):
    product_kind: PriceDeskKind
    price_list_date: str | None = None
    file_sha256: str
    parsed_rows: int
    changed: int
    new: int
    missing: int
    unchanged: int
    examples: PriceDeskDiffExamples


class PriceGroupStatus(BaseModel):
    product_kind: PriceDeskKind
    price_list_date: str | None = None
    imported_at: str | None = None
    row_count: int = 0


class PriceDeskStatus(BaseModel):
    groups: list[PriceGroupStatus]

File: frontend/src/features/price-desk/types/priceDesk.ts
/** Типы API стола прайсов — синхронны с app/schemas/price_desk.py. */

export const PRICE_DESK_KINDS = [
  "plates",
  "fbs",
  "march",
  "step",
  "bridge_pile",
  "pile",
] as const;

export type PriceDeskKind = (typeof PRICE_DESK_KINDS)[number];

export const PRICE_DESK_KIND_LABELS: Record<PriceDeskKind, string> = {
  plates: "Плиты ПБ",
  fbs: "ФБС",
  march: "ЛМ",
  step: "Ступени",
  bridge_pile: "Мостовые сваи",
  pile: "Цельные сваи",
};

export type PriceDeskDiffExample = {
  key: string;
  old_price: number | null;
  new_price: number | null;
};

export type PriceDeskDiffExamples = {
  changed: PriceDeskDiffExample[];
  new: PriceDeskDiffExample[];
  missing: PriceDeskDiffExample[];
  unchanged: PriceDeskDiffExample[];
};

export type PriceDeskPreview = {
  product_kind: PriceDeskKind;
  price_list_date: string | null;
  file_sha256: string;
  parsed_rows: number;
  changed: number;
  new: number;
  missing: number;
  unchanged: number;
  examples: PriceDeskDiffExamples;
};

export type PriceGroupStatus = {
  product_kind: PriceDeskKind;
  price_list_date: string | null;
  imported_at: string | null;
  row_count: number;
};

export type PriceDeskStatus = {
  groups: PriceGroupStatus[];
};

export const priceDeskKindLabel = (kind: string): string => {
  if (kind in PRICE_DESK_KIND_LABELS) {
    return PRICE_DESK_KIND_LABELS[kind as PriceDeskKind];
  }
  return kind;
};
